### PR TITLE
Anonymizer Hash

### PR DESCRIPTION
**Pull request by balteravishay**
Add support for anonymization using hash.
Allows hashing using either SHA256, SHA512, MD5. default is SHA256.
includes doc (anonymizer MD), unit tests and functional tests

### CODE REVIEW

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py [hunk 1]
```diff
@@ -2,8 +2,27 @@
 from presidio_anonymizer.entities import InvalidParamException
 
 
+def validate_parameter_in_range(
+    values_range, parameter_value, parameter_name: str, parameter_type: type
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:6]: you don't use parameter_type

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py [hunk 2]
```diff
@@ -2,8 +2,27 @@
...
+) -> None:
+    """Validate an anonymizer parameter.
+
+    valdidates the existence of an anonymizer parameter and that it is an
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:10]: typo in valdidates

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py [hunk 3]
```diff
@@ -2,8 +2,27 @@
...
+    Otherwise, raise the appropriate InvalidParamException with the
+    parameter_name as content.
+    """
+    validate_parameter(parameter_value, parameter_name, parameter_type)
+    if parameter_value not in values_range:
+        raise InvalidParamException(
+            f"Parameter {parameter_name} value {parameter_value} is not in "
+            f"range of values {values_range}"
+        )
+    pass
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:21]: intentional?

File: presidio-anonymizer/README.MD [hunk 4]
```diff
@@ -13,7 +13,9 @@ Presidio anonymizer comes by default with the following anonymizers:
   If "new_value" is not supplied or empty, default behavior will be: <entity_type>
   e.g: <PHONE_NUMBER>
 - Redact - removes the PII completely from text Parameters: None
-- Hash - hash the PII using sha256
+- Hash - hash the PII using either sha256, sha512 or md5.
+  Parameters:
+  * "hash_type" - sets the type of hashing. can be either sha256, sha512 or md5.
```
Comment (SharonHart) [inline presidio-anonymizer/README.MD:18]: Add that we're using sha256 by default if not passed

File: presidio-anonymizer/tests/anonymizers/test_mask.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
         ("t", "*", 3, False, "*"),  # with overflowing chars_to_mask
         ("t", "*", 3, True, "*"),  # from the end
         ("text", "😈", 4, False, "😈😈😈😈"),  # Mask with 'Unicode EmojiSources' character
+        ("😈😈😈😈", "*", 4, False, "****"),  # Mask 'Unicode EmojiSources' character
```
Comment (SharonHart) [inline presidio-anonymizer/tests/anonymizers/test_mask.py:22]: 👍

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py [hunk 6]
```diff
@@ -1,19 +1,37 @@
...
+            self.MD5: lambda s: md5(s),
+        }
+        return hash_switcher.get(hash_type)(text.encode()).hexdigest()
 
     def validate(self, params: dict = None) -> None:
-        """TODO: [ADO-2544] docstring."""
+        """Validate the hash type is string and in range of allowed hash types."""
+        validate_parameter_in_range(
+            [self.SHA256, self.SHA512, self.MD5],
+            params.get(self.HASH_TYPE, self.SHA256),
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py:33]: we're defining the default twice, one might forget to update both if changed.
Either allow self.HASH_TYPE to be empty on validate, check that only if exists then validate, or define a default and use both in anonymize and validate

File: presidio-anonymizer/tests/anonymizers/test_hash.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import pytest
+
+from presidio_anonymizer.anonymizers import Hash
+
```
Comment (SharonHart) [inline presidio-anonymizer/tests/anonymizers/test_hash.py:4]: What about validate() tests?

File: presidio-anonymizer/tests/anonymizers/test_hash.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        ),  # Sha512  Hash 'Unicode EmojiSources' character
+        ("123456", "md5", "e10adc3949ba59abbe56e057f20f883e"),  # MD5 Hash 123456
+        ("54321", "md5", "01cfcd4f6b8770febfb40cb906715822"),  # MD5 Hash 54321
+        (
+            "😈😈😈😈",
+            "md5",
+            "5bf45eeaade2060ac6cdd532e1c35eef",
+        ),  # MD5 Hash 'Unicode EmojiSources' character
+    ]
+    # fmt: on
```
Comment (SharonHart) [inline presidio-anonymizer/tests/anonymizers/test_hash.py:77]: you're missing the off, was it enough for disabling autoformat?

File: presidio-anonymizer/tests/anonymizers/test_validators.py [hunk 9]
```diff
@@ -1,10 +1,33 @@
...
+    ):
+        validate_parameter_in_range(
+            values_range=["0", "2"],
+            parameter_value="1",
+            parameter_name="name",
+            parameter_type=str,
+        )
+
+
+def test_when_parameter_in_range_we_pass():
```
Comment (SharonHart) [inline presidio-anonymizer/tests/anonymizers/test_validators.py:22]: then_we_pass

File: functional-tests/tests/test_anonymizer.py [hunk 10]
```diff
@@ -29,6 +29,33 @@ def test_given_anonymize_called_with_valid_request_then_expected_valid_response_
     assert response_content == expected_response
 
 
+@pytest.mark.api
+def test_anonymize_hash_called_with_valid_request_then_expected_valid_response_returned():
```
Comment (SharonHart) [inline functional-tests/tests/test_anonymizer.py]: test_when_anonymize

File: presidio-anonymizer/tests/anonymizers/test_validators.py [hunk 11]
```diff
@@ -34,22 +57,13 @@ def test_when_actual_parameter_is_non_json_typed_then_ipe_raised_with_general_er
...
     err_str = "Invalid parameter value for name. Expecting 'string', but got 'number'."
-    with pytest.raises(
-            InvalidParamException,
-            match=err_str
-    ):
-        validate_type(
-            parameter_value=1, parameter_name="name", parameter_type=str
-        )
+    with pytest.raises(InvalidParamException, match=err_str):
+        validate_type(parameter_value=1, parameter_name="name", parameter_type=str)
```
Comment (SharonHart) [inline presidio-anonymizer/tests/anonymizers/test_validators.py:61]: black autoformatted it for me, so in case we'll change this file again it will make the format as it was again. We can have discussion on using black again, or try to refine it somehow since it gives some unreadable results sometimes

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py (hunk 1 above)
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:6]: do we need the type if we use range?

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py [hunk 12]
```diff
@@ -1,19 +1,37 @@
...
+    MD5 = "md5"
 
     def anonymize(self, text: str = None, params: dict = None) -> str:
         """
         Hash given value using sha256.
 
         :return: hashed original text
         """
-        return params.get("old_text")
+        hash_type = params.get(self.HASH_TYPE, self.SHA256)
```
Comment (navalev) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py:21]: are we documenting that the default is sha265?

File: presidio-anonymizer/tests/anonymizers/test_hash.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+)
+def test_when_given_valid_value_with_hash_type_then_expected_string_returned(
```
Comment (navalev) [inline presidio-anonymizer/tests/anonymizers/test_hash.py:79]: lets also add a test with empty string

**Review by navalev — COMMENTED** [review]
looks good, left a couple of comments. once all resolved will approve

File: presidio-anonymizer/README.MD [hunk 14]
```diff
@@ -13,7 +13,9 @@ Presidio anonymizer comes by default with the following anonymizers:
   If "new_value" is not supplied or empty, default behavior will be: <entity_type>
   e.g: <PHONE_NUMBER>
 - Redact - removes the PII completely from text Parameters: None
-- Hash - hash the PII using sha256
+- Hash - hash the PII using either sha256, sha512 or md5.
```
Comment (shiranr) [inline presidio-anonymizer/README.MD:16]: Very nice addition! I don't think we had it in the previous anonymizer!

File: presidio-anonymizer/tests/anonymizers/test_mask.py (hunk 5 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_mask.py:22]: :)

File: presidio-anonymizer/tests/anonymizers/test_validators.py (hunk 11 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_validators.py:61]: so what's to do here?

Comment (SharonHart) [reply presidio-anonymizer/tests/anonymizers/test_validators.py:61]: What did the reformatting for you? IDE?

File: presidio-anonymizer/README.MD (hunk 14 above)
Comment (balteravishay) [reply presidio-anonymizer/README.MD:16]: thanks. resolve :)

File: presidio-anonymizer/README.MD (hunk 4 above)
Comment (balteravishay) [reply presidio-anonymizer/README.MD:18]: done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py (hunk 12 above)
Comment (balteravishay) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py]: done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py (hunk 6 above)
Comment (balteravishay) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py]: done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py (hunk 2 above)
Comment (balteravishay) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py]: done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py (hunk 1 above)
Comment (balteravishay) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:6]: done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py (hunk 3 above)
Comment (balteravishay) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py]: done

File: presidio-anonymizer/tests/anonymizers/test_hash.py (hunk 7 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_hash.py:5]: done

File: presidio-anonymizer/tests/anonymizers/test_hash.py (hunk 8 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_hash.py]: done

File: presidio-anonymizer/tests/anonymizers/test_hash.py (hunk 13 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_hash.py:84]: done

File: presidio-anonymizer/tests/anonymizers/test_validators.py (hunk 9 above)
Comment (balteravishay) [reply presidio-anonymizer/tests/anonymizers/test_validators.py]: done

**Review by navalev — APPROVED** [review]

File: presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py (hunk 1 above)
Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/validators.py:6]: I meant maybe not pass parameter_type at all?

File: functional-tests/Pipfile [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+name = "pypi"
+
+[packages]
+requests = "*"
+pytest = "*"
+
+[dev-packages]
+
+[requires]
+python_version = "3.8"
```
Comment (SharonHart) [inline functional-tests/Pipfile:13]: Got here by mistake?

**Review by SharonHart — CHANGES_REQUESTED** [review]

File: functional-tests/Pipfile (hunk 15 above)
Comment (balteravishay) [reply functional-tests/Pipfile:13]: yes, good catch

**Review by navalev — APPROVED** [review]

**Review by SharonHart — APPROVED** [review]
